### PR TITLE
range: fix overflow value access index 

### PR DESCRIPTION
**Pull request by luxinle**
### What problem does this PR solve?

related pr #21805 #17253

Problem Summary:

### What is changed and how it works?

What's Changed:
When the value overflows access index. Replace with a fixed constant value.
How it Works:

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- N/A
### Release note 

- No release note

### CODE REVIEW

**Review by xuyifangreeneyes — COMMENTED** [review]
LGTM

File: util/ranger/points.go [hunk 1]
```diff
@@ -350,34 +350,32 @@ func handleUnsignedCol(ft *types.FieldType, val types.Datum, op string) (types.D
...
 		return val, op, true
 	}
-
 	switch ft.Tp {
 	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong:
-		if !isNegative && val.GetUint64() > math.MaxInt64 {
+		if val.Kind() == types.KindUint64 && val.GetUint64() > math.MaxInt64 {
 			switch op {
-			case ast.GT, ast.GE:
+			case ast.GT, ast.GE, ast.EQ, ast.NullEQ:
```
Comment (Reminiscent) [inline util/ranger/points.go:360]: Please add more tests for this.

File: util/ranger/points.go [hunk 2]
```diff
@@ -350,34 +350,32 @@ func handleUnsignedCol(ft *types.FieldType, val types.Datum, op string) (types.D
...
 				op = ast.LE
 				val = types.NewIntDatum(math.MaxInt64)
 			}
 		}
 	case mysql.TypeFloat:
-		if val.GetFloat64() > math.MaxFloat32 {
+		if val.Kind() == types.KindFloat64 && val.GetFloat64() > math.MaxFloat32 {
 			switch op {
-			case ast.GT, ast.GE:
+			case ast.GT, ast.GE, ast.EQ, ast.NullEQ:
```
Comment (Reminiscent) [inline util/ranger/points.go:370]: ditto

File: util/ranger/points.go [hunk 3]
```diff
@@ -350,34 +350,32 @@ func handleUnsignedCol(ft *types.FieldType, val types.Datum, op string) (types.D
...
 				return val, op, false
 			case ast.NE, ast.LE, ast.LT:
 				op = ast.LE
 				val = types.NewFloat32Datum(math.MaxFloat32)
 			}
-		} else if val.GetFloat64() < -math.MaxFloat32 {
+		} else if val.Kind() == types.KindFloat64 && val.GetFloat64() < -math.MaxFloat32 {
 			switch op {
-			case ast.LE, ast.LT:
+			case ast.LE, ast.LT, ast.EQ, ast.NullEQ:
```
Comment (Reminiscent) [inline util/ranger/points.go:378]: ditto

**Comment by xuyifangreeneyes** [comment]
@luxinle is there any update?

File: util/ranger/ranger_test.go [hunk 4]
```diff
@@ -826,21 +861,26 @@ create table t(
...
+			res, err := ranger.DetachCondAndBuildRangeForIndex(sctx, conds, cols, lengths)
+			c.Assert(err, IsNil)
+			c.Assert(fmt.Sprintf("%s", res.AccessConds), Equals, tt.accessConds, Commentf("wrong access conditions for expr: %s", tt.exprStr))
+			c.Assert(fmt.Sprintf("%s", res.RemainedConds), Equals, tt.filterConds, Commentf("wrong filter conditions for expr: %s", tt.exprStr))
+			got := fmt.Sprintf("%v", res.Ranges)
+			c.Assert(got, Equals, tt.resultStr, Commentf("different for expr %s", tt.exprStr))
+		}
+		if selection, ok := p.(plannercore.LogicalPlan).Children()[0].(*plannercore.LogicalTableDual); ok {
+			c.Assert(selection, NotNil, Commentf("expr:%v", tt.exprStr))
+			continue
```
Comment (xuyifangreeneyes) [inline util/ranger/ranger_test.go:882]: Which test case would enter here?

Comment (luxinle) [reply util/ranger/ranger_test.go:882]: col_int <=> 9223372036854775808  or col_int = 9223372036854775808

File: util/ranger/ranger_test.go [hunk 5]
```diff
@@ -812,6 +812,34 @@ create table t(
...
+		{
+			indexPos:    5,
+			exprStr:     "col_float <=> 1000000000000000000000000000000000000000",
+			accessConds: "[nulleq(test.t.col_float, 1e+39)]",
+			filterConds: "[]",
+			resultStr:   "[]",
+		},
+		{
+			indexPos:    5,
+			exprStr:     "col_float = -1000000000000000000000000000000000000000",
```
Comment (zz-jason) [inline util/ranger/ranger_test.go:831]: I think we have fixed this, check this in the master branch:

```sql
create table t(a float);
explain select * from t where a > -1000000000000000000000000000000000000000;
```

the result in master branch is:

```sql
TiDB(root@127.0.0.1:test) > explain select * from t where a > -1000000000000000000000000000000000000000;
+-------------------------+----------+-----------+---------------+--------------------------------+
| id                      | estRows  | task      | access object | operator info                  |
+-------------------------+----------+-----------+---------------+--------------------------------+
| TableReader_7           | 3333.33  | root      |               | data:Selection_6               |
| └─Selection_6           | 3333.33  | cop[tikv] |               | gt(test.t.a, -1e+39)           |
|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+---------------+--------------------------------+
3 rows in set (0.00 sec)
```

which is the same this test case.

**Comment by zz-jason** [comment]
@luxinle could you explain more about what issue is this PR aimed to fix?

**Comment by luxinle** [comment]
The master code already fixed this.

**Comment by luxinle** [comment]
close this pr

**Pull request closed without merging**